### PR TITLE
#10271 t.w.error.Error: Fix handling of `message=None`

### LINKED ISSUE #10271: Problem in web.error str conversion
|[technic's avatar](https://github.com/technic)| @technic reported|
|-|-|
|Trac ID|trac#10271|
|Type|defect|
|Created|2021-10-31 22:06:47Z|

Hi,

This method returns an error

https://github.com/twisted/twisted/blob/3d0be31827585ade5078624e17a1c15240eece02/src/twisted/web/error.py#L92

"can't concat str and None" because self.message is None.

in __init__() message is empty string, and the code is 520, then _codeToMessage() returns None, and finally self.message is None.

I suggest to change the code to something like `self.message = message or "Unknown"`

<details><summary>Searchable metadata</summary>

```
trac-id__10271 10271
type__defect defect
reporter__technic technic
priority__normal normal
milestone__None None
branch__ 
branch_author__ 
status__new new
resolution__None None
component__web web
keywords__None None
time__1635718007391402 1635718007391402
changetime__1635718007391402 1635718007391402
version__None None
owner__None None

```
</details>

### PR DESCRIPTION
**Pull request by twm**
Also add type annotations and validate HTTP status codes.

## Scope and purpose

Fixes #10271.

## Contributor Checklist:

This process applies to *all* pull requests - no matter how small.
Have a look at [our developer documentation](https://docs.twisted.org/en/latest/core/development/dev-process.html) before submitting your Pull Request.

Below is a non-exhaustive list (as a reminder):

* The title of the PR should describe the changes and starts with the associated issue number, like “#1234 Brief description”.
* A release notes news fragment file was create in src/twisted/newsfragments/ (see: [Release notes fragments docs.](https://docs.twisted.org/en/latest/core/development/dev-process.html#release-notes-management))
* The automated tests were updated.
* Once all checks are green, request a review by leaving a comment that contains exactly the string `please review`.
  Our bot will trigger the review process, by applying the pending review label
  and requesting a review from the Twisted dev team.

### CODE REVIEW

**Comment by twm** [comment]
Could we stop changing the PR title format? :roll_eyes:

**Comment by twm** [comment]
please review

**Comment by adiroiban** [comment]
The PR title was updates to make it easier to navigate the history.

Now, if you look at the commit history, you get quick links to both the issue and the PR

image


sorry for the trouble.

File: src/twisted/web/newsfragments/10271.bugfix
Comment (adiroiban) [inline src/twisted/web/newsfragments/10271.bugfix:1]: The current text is ok.

I am just requestion some feedback for towncrier.

I think there are times when you want 2 independent bugfixes for a single PR 

It would be better to show them in the final release as separate entries.

Rigth now, towcrier wraps everything that it find in the fragment file

Maybe something like this (with en empty line) is better and can instruct towncrier to generate 2 separate event handlers.

```suggestion
twisted.web.error.Error.__str__ no longer raises an exception when the error's message attribute is None.

twisted.web.error.Error validates at initialization that code is a plausible 3-digit HTTP status code.
```

File: src/twisted/web/test/test_error.py
Comment (adiroiban) [inline src/twisted/web/test/test_error.py:67]: Maybe something like this, to also check that we don't get a any ValueError

```suggestion
    error = self.assertRaises(ValueError, error.Error, b"InvalidStatus")
    self.assertEqual("Not a valid HTTP status code: InvalidStatus", error.args[0])
```

**Review by adiroiban — APPROVED** [review]
This looks good. Thanks for the cleanup.

But why do we allow message=None in the first place?

I would say that if we pass message=None at init, `web.error.Error` should automatically resolve it into some sort of default message.

That is, call `_codeToMessage` right away.

In this way, later, when you have to deal with the error, you don't need to create all kinds of special cases for message=None

But that should not block this PR..and maybe such a change would be a backward incompatible change.

File: src/twisted/web/error.py
Comment (adiroiban) [inline src/twisted/web/error.py:31]: this is private method...so we should be fine.

But maybe is better to be strict and only allow bytes as input?

File: src/twisted/web/error.py
Comment (adiroiban) [inline src/twisted/web/error.py:122]: this is already in trunk, but does it make sense to have a redirection without a location ? Is it useful ?

File: src/twisted/web/test/test_error.py
Comment (adiroiban) [reply src/twisted/web/test/test_error.py:67]: I am -1 for using the context manager for `assertRaises` as it looks there are people who will add assertions inside the context manager, and will think that all is good, when in fact the assertion code is never executed.

I hope that the code coverage can help identity these types of errors, so I don't think we should enforce this.

is just a nice to have.

File: src/twisted/web/newsfragments/10271.bugfix
Comment (twm) [reply src/twisted/web/newsfragments/10271.bugfix:1]: In this case I think that the way to split this is to create a 10271.feature in addition to the .bugfix.

I am generally skeptical of the notion that we should work out these details at the PR level. IMO revising the combined release notes should be done as part of the release process by a human being. There is no way we can extract consistent prose from a bunch of fragment files written by random contributors.

File: src/twisted/web/error.py
Comment (twm) [reply src/twisted/web/error.py:122]: I think we issue this error based on status code without regard to whether there is actually a Location header. Is this good? 🤷‍♂️

File: src/twisted/web/error.py
Comment (twm) [reply src/twisted/web/error.py:31]: I did that in an earlier draft, but it led to awkward control flow to avoid coercing things that aren't already coerced (the `super()` arguments). I didn't think the compatibility implications of this worth the tidings of an internal interface.

File: src/twisted/web/test/test_error.py
Comment (twm) [reply src/twisted/web/test/test_error.py:67]: I appreciate this but as the suggested change is not correct and I'm not at my workstation I'm going to lea e it as-is.

File: src/twisted/web/newsfragments/10271.bugfix
Comment (adiroiban) [reply src/twisted/web/newsfragments/10271.bugfix:1]: well... we already do just a few releases per year and don't have many people volunteering for doing the release.

This is why I try to make the release process as simple and as quick as possible.

So making sure that random contributors are writing good release notes from the start, is one step in reducing the burden on the release manager :p

File: src/twisted/web/error.py
Comment (adiroiban) [reply src/twisted/web/error.py:122]: I don't know... I guess is ok... but not ideal.

A redirect without location, looks to me like a broken response :)

File: src/twisted/web/error.py
Comment (DMRobertson) [inline src/twisted/web/error.py:65]: I believe this change makes `PartialDownloadError.response: Optional[bytes]`. Is that correct? If we've got a partial download, we probably have some bytes? I suppose those might be all header bytes rather than response body bytes though, in which case maybe `response: None` does make sense?

(Noticed by https://github.com/matrix-org/synapse/issues/14008)

Comment (DMRobertson) [reply src/twisted/web/error.py:65]: FWIW, The only call site I could find is

https://github.com/twisted/twisted/blob/ab59aae25cd18c2cda96b16c165900cfc57f1f38/src/twisted/web/client.py#L1718-L1720

which gives a non-None `response` object.

Comment (twm) [reply src/twisted/web/error.py:65]: Seems likely. You could submit a ticket and PR to refine it in `PartialDownloadError` I guess?

**Pull request merged**
